uswds/uswds
# USWDS - Settings: Update sass map pattern

# Summary

Updates sass map pattern to allow users to only define overrides while keeping defaults for the other values

## Breaking change

This is _potentially_ a breaking change.

## Related issue

Closes #5215

## Related pull requests

Update sass map instructions on utility pages: https://github.com/uswds/uswds-site/pull/2077

## Problem statement

Currently, if you choose to update a USWDS setting within a sass map, you must define each setting within the map even if you are keeping the default values for the other.

## Solution

We can use map-merge with a default values map to only replace updated settings keys without needing to redefine default values.

This solution
- Doesn't break existing sass map settings users have previously set
- _BUT_ if users left an element out of the map, (effectively disabling it) this PR would re-enable the default setting.

This PR updates `$theme-[map-name]`s to `$theme-[map-name]-complete` wherever the setting is called.

This allows users still use existing setting names, such as `$theme-utility-breakpoints`, to update which breakpoints are active, while using the `-complete` map to restore defaults to undefined elements.

> Note: I considered naming the `complete` maps as `-defaults` but I believe this makes less contextual sense wherever the maps are being called.

## Updated maps

- [x] `$theme-namespace`
- [x] `$theme-utility-breakpoints`
- [x] `$[utility-module]-settings` 

## Ignored maps

These maps were either, empty, set to `false` by default, or only had example values. In all three settings, we would want the users override to be the only value, which is currently how they are set up.

- `$theme-fontface-tokens`
- `$theme-font-[family]-src`
- `$[utility-module]-manual-calues`

## Testing and review

1. Checkout this [test uswds-sandbox repo](https://github.com/uswds/uswds-sandbox/tree/cm-breakpoint-map)
3. The test repo should already have this branch installed as it's USWDS package
4. Run `npm run start` on the test repo
5. View each test page and review instructions
7. Visit `_uswds-theme.scss` and expirament enabling/disabling different settings
8. Revisit index page in browser and view the changes
9. Confirm that
   - [ ] : Default breakpoints are working as expected
   - [ ] : Updating one breakpoint, doesn't deactivate the others
   - [ ] : Updated breakpoints work as expected
10. Look over all maps listed on the [settings](https://designsystem.digital.gov/documentation/settings/#general-settings-2) page
11. Inspect their default values in the USWDS settings and verify there aren't any additional updates to make

These tests are simple and mostly just enable different background colors as a quick visual feedback. Feel free to experiment with other utility settings to verify the new pattern is working!

### Default maps for reference:
 
**Breakpoints**

```scss
@use "uswds-core" with (
  $theme-utility-breakpoints: (
      "card": false,
      "card-lg": false,
      "mobile": false,
      "mobile-lg": true,
      "tablet": true,
      "tablet-lg": false,
      "desktop": true,
      "desktop-lg": false,
      "widescreen": false,
    )
);
```

**Namespace**
```scss
$theme-namespace: (
  "grid": (
    namespace: "grid-",
    output: true
  ),
  "utility": (
    namespace: "u-",
    output: false,
  ),
),
```

Before opening this PR, make sure you’ve done whichever of these applies to you:
